python/utilities/dnslookup/dnslookup.py

```diff
@@ -0,0 +1,130 @@
+#!/usr/bin/env python
+
+# Implement DNS inclusion proof checking, see [TBD].
+#
+# Unfortunately, getting at the SCTs in general is hard in Python, so this
+# does not start with an SSL connection, but instead fetches a log entry by
+# index and then verifies the proof over DNS.
+
+import base64
+import dns.resolver
+import hashlib
+import json
+import logging
+import os
+import subprocess
+import sys
+import tempfile
+import urllib2
+
+ALL_LOGS_URL = 'https://www.certificate-transparency.org/known-logs/all_logs_list.json'
+
+basepath = os.path.dirname(sys.argv[0])
+sys.path.append(os.path.join(basepath, '../../../python'))
+from ct.crypto import merkle, verify
+from ct.proto import ct_pb2, client_pb2
+
+ctpath = os.path.join(basepath, '../../../cpp/client/ct')
+libpath = os.path.join(basepath, '../../../../install/lib')
+
+class CTDNSLookup:
+    def __init__(self, domain, verifier, nameservers, port = 53):
+        self.verifier = verifier
+        self.domain = domain
+        self.resolver = dns.resolver.Resolver(configure=False)
+        self.resolver.nameservers = nameservers
+        self.resolver.port = port
+
+    def Get(self, name):
+        logging.info("get " + name)
+        answers = self.resolver.query(name, 'TXT')
+        assert answers.rdtype == dns.rdatatype.TXT
+        return answers
+
+    def GetOne(self, name):
+        name += '.' + self.domain
+        answers = self.Get(name)
+        assert len(answers) == 1
+        txt = answers[0]
+        assert len(txt.strings) == 1
+        return txt.strings[0]
+
+    def GetSTH(self):
+        sth_str = self.GetOne('sth')
+        sth = client_pb2.SthResponse()
+        parts = str(sth_str).split('.')
+        sth.tree_size = int(parts[0])
+        sth.timestamp = int(parts[1])
+        sth.sha256_root_hash = base64.b64decode(parts[2])
+        sth.tree_head_signature = base64.b64decode(parts[3])
+
+        self.verifier.verify_sth(sth)
+
+        return sth
+
+    def GetEntry(self, level, index, size):
+        return self.GetOne(str(level) + '.' + str(index) + '.' + str(size)
+                           + '.tree')
+
+    def GetIndexFromHash(self, hash):
+        return self.GetOne(base64.b32encode(hash)[:-4] + '.hash')
```

Comment:
This `-4` here depends on the the hash algorithm, maybe use a constant for that? Not going to have you make it compute the padding from the length of the hash output... :wink: